File: src/main/java/br/com/jiratorio/mapper/BoardMapper.java
package br.com.jiratorio.mapper;

import br.com.jiratorio.domain.entity.Board;
import br.com.jiratorio.domain.request.CreateBoardRequest;
import br.com.jiratorio.domain.request.UpdateBoardRequest;
import br.com.jiratorio.domain.response.BoardDetailsResponse;
import br.com.jiratorio.domain.response.BoardResponse;
import br.com.jiratorio.mapper.transformer.StringTransformer;
import java.util.List;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.MappingTarget;
import org.mapstruct.Mappings;
import org.mapstruct.ReportingPolicy;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;

@Mapper(componentModel = "spring",
        unmappedTargetPolicy = ReportingPolicy.IGNORE,
        uses = StringTransformer.class)
public interface BoardMapper {

    BoardDetailsResponse toBoardResponseDetails(Board board);

    Board boardFromCreateBoardRequest(CreateBoardRequest request);

    List<BoardResponse> toBoardResponse(List<Board> boards);

    default Page<BoardResponse> toBoardResponse(Page<Board> boards) {
        List<BoardResponse> boardResponses = toBoardResponse(boards.getContent());
        return new PageImpl<>(boardResponses, boards.getPageable(), boards.getTotalElements());
    }

    @Mappings({ // @formatter:off
        @Mapping(target = "startColumn",     qualifiedByName = "toUpperCase"),
        @Mapping(target = "endColumn",       qualifiedByName = "toUpperCase"),
        @Mapping(target = "fluxColumn",      qualifiedByName = "listToUpperCase"),
        @Mapping(target = "touchingColumns", qualifiedByName = "listToUpperCase"),
        @Mapping(target = "waitingColumns",  qualifiedByName = "listToUpperCase")
    }) // @formatter:on
    void fromUpdateBoardRequest(@MappingTarget Board board, UpdateBoardRequest updateBoardRequest);
}

File: src/main/java/br/com/jiratorio/mapper/BoardMapper.kt
package br.com.jiratorio.mapper

import br.com.jiratorio.domain.entity.Board
import br.com.jiratorio.domain.request.CreateBoardRequest
import br.com.jiratorio.domain.request.UpdateBoardRequest
import br.com.jiratorio.domain.response.BoardDetailsResponse
import br.com.jiratorio.domain.response.BoardResponse
import br.com.jiratorio.mapper.transformer.StringTransformer
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageImpl
import org.springframework.stereotype.Component

@Component
class BoardMapper(
    private val stringTransformer: StringTransformer
) {

    fun toBoardResponseDetails(board: Board) =
        BoardDetailsResponse(
            externalId = board.externalId,
            name = board.name,
            startColumn = board.startColumn,
            endColumn = board.endColumn,
            fluxColumn = board.fluxColumn,
            ignoreIssueType = board.ignoreIssueType,
            epicCF = board.epicCF,
            estimateCF = board.estimateCF,
            systemCF = board.systemCF,
            projectCF = board.projectCF,
            ignoreWeekend = board.ignoreWeekend,
            impedimentType = board.impedimentType,
            impedimentColumns = board.impedimentColumns,
            dynamicFields = board.dynamicFields,
            dueDateCF = board.dueDateCF,
            dueDateType = board.dueDateType?.name
        )

    fun boardFromCreateBoardRequest(request: CreateBoardRequest) =
        Board(
            name = request.name,
            externalId = request.externalId
        )

    fun toBoardResponse(board: Board) = BoardResponse(
        id = board.id,
        name = board.name,
        owner = board.owner
    )

    fun toBoardResponse(boards: List<Board>) =
        boards.map { toBoardResponse(it) }

    fun toBoardResponse(boards: Page<Board>) =
        PageImpl(
            toBoardResponse(boards.content),
            boards.pageable,
            boards.totalElements
        )

    fun fromUpdateBoardRequest(board: Board, updateBoardRequest: UpdateBoardRequest) =
        board.apply {
            name = updateBoardRequest.name
            startColumn = stringTransformer.toUpperCase(updateBoardRequest.startColumn)
            endColumn = stringTransformer.toUpperCase(updateBoardRequest.endColumn)
            fluxColumn = stringTransformer.listToUpperCase(updateBoardRequest.fluxColumn)
            ignoreIssueType = updateBoardRequest.ignoreIssueType
            epicCF = updateBoardRequest.epicCF
            estimateCF = updateBoardRequest.estimateCF
            systemCF = updateBoardRequest.systemCF
            projectCF = updateBoardRequest.projectCF
            ignoreWeekend = updateBoardRequest.ignoreWeekend
            impedimentType = updateBoardRequest.impedimentType
            impedimentColumns = stringTransformer.listToUpperCase(updateBoardRequest.impedimentColumns)
            dynamicFields = updateBoardRequest.dynamicFields
            touchingColumns = stringTransformer.listToUpperCase(updateBoardRequest.touchingColumns)
            waitingColumns = stringTransformer.listToUpperCase(updateBoardRequest.waitingColumns)
            dueDateCF = updateBoardRequest.dueDateCF
            dueDateType = updateBoardRequest.dueDateType
        }
}
